core/drand_test.go

```diff
@@ -37,228 +32,324 @@ func setFDLimit() {
+
+	t.Log("[reshare] Move to response phase!")
+	t.Logf("[reshare] Group: %s", group3)
 }
 
 // This tests when a node first signal his intention to participate into a
 // resharing but is down right after  - he shouldn't be in the final group
-func TestDrandDKGReshareAbsent(t *testing.T) {
-	oldN := 3
-	newN := 4
-	oldThr := 2
-	newThr := 3
-	timeout := 1 * time.Second
-	beaconPeriod := 2 * time.Second
+func TestRunDKGReshareAbsentNode(t *testing.T) {
+	oldNodes, newNodes := 3, 4
+	oldThreshold, newThreshold := 2, 3
+	timeout, beaconPeriod := 1*time.Second, 2*time.Second
 
-	dt := NewDrandTest2(t, oldN, oldThr, beaconPeriod)
+	dt := NewDrandTestScenario(t, oldNodes, oldThreshold, beaconPeriod)
 	defer dt.Cleanup()
+
 	group1 := dt.RunDKG()
-	// make sure all nodes had enough time to run their go routines to start the
-	// beacon handler - related to CI problems
+
+	// TODO: make sure all nodes had enough time to run their go routines to start the beacon handler - related to CI problems
 	time.Sleep(getSleepDuration())
-	dt.MoveToTime(group1.GenesisTime)
+	dt.SetMockClock(t, group1.GenesisTime)
+
 	// move to genesis time - so nodes start to make a round
-	// dt.MoveTime(offsetGenesis)
+	// dt.AdvanceMockClock(t,offsetGenesis)
 	// two = genesis + 1st round (happens at genesis)
-	dt.TestBeaconLength(2, false, dt.Ids(oldN, false)...)
+
+	t.Log("Check Beacon Length")
+	dt.CheckBeaconLength(2, false, dt.Ids(oldNodes, false)...)
+
 	// so nodes think they are going forward with round 2
-	dt.MoveTime(1 * time.Second)
+	dt.AdvanceMockClock(t, 1*time.Second)
+
+	t.Log("Adding new nodes to the group")
+	nodesToAdd := newNodes - oldNodes
+	dt.SetupNewNodes(t, nodesToAdd)
 
-	toAdd := newN - oldN
-	dt.SetupNewNodes(toAdd)
 	// we want to stop one node right after the group is created
 	nodeToStop := 1
 	leader := 0
+
+	t.Log("Stopping nodes\n")
 	dt.nodes[leader].drand.setupCB = func(g *key.Group) {
-		fmt.Printf("\n\nSTOPPING NODE 1\n\n")
+		t.Logf("Stopping node %d \n", nodeToStop)
 		dt.nodes[nodeToStop].drand.Stop(context.Background())
-		fmt.Printf("\n\nSTOPPING NODE 1 DONE \n\n")
+		t.Logf("Node %d stopped \n", nodeToStop)
 	}
-	fmt.Println("SETUP RESHARE DONE")
-	newGroup, err := dt.RunReshare(oldN, toAdd, newThr, timeout, false, false, true)
+
+	t.Log("Setup reshare done. Starting reshare... Ignoring reshare errors")
+	newGroup, err := dt.RunReshare(t, nil, oldNodes, nodesToAdd, newThreshold, timeout, false, false, true)
 	require.NoError(t, err)
 	require.NotNil(t, newGroup)
+
 	// the node that had stopped must not be in the group
+	t.Logf("Check node %d is not included in the group \n", nodeToStop)
 	missingPublic := dt.nodes[nodeToStop].drand.priv.Public
 	require.Nil(t, newGroup.Find(missingPublic), "missing public is found", missingPublic)
 }
 
-func TestDrandDKGReshareTimeout(t *testing.T) {
-	oldN := 3
-	newN := 4
-	oldThr := 2
-	newThr := 3
-	timeout := 1 * time.Second
-	beaconPeriod := 2 * time.Second
+// Test ??????????????????????
```

Comment:
This test creates the scenario where one node made a complaint during the DKG, at the second phase, so normally, there should be a "Justification" at the third phase. In this case, there is not. This scenario can happen if there is an offline node right at the beginning of DKG that don't even send any message.

Reply:
Perfect!